src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,307 @@
+            // Namespaces do not have attributes
+            while (containingType is INamespaceSymbol)
+            {
+                containingType = containingType.ContainingType;
+            }
+            if (containingType != null)
+            {
+                return ProcessPreviewAttribute(containingType, requiresPreviewFeaturesSymbols);
+            }
+
+            return false;
+        }
+
+        private static bool SymbolIsStaticAndAbstract(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            bool isStaticAndAbstract = symbol.IsStatic && symbol.IsAbstract;
+
+            return isStaticAndAbstract && !ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static bool PropertyOrMethodSymbolHasAttribute(ISymbol propertyOrMethodSymbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
+        {
+            if (SymbolIsStaticAndAbstract(propertyOrMethodSymbol, requiresPreviewFeaturesSymbols))
+            {
+                // TODO: This needs to be tagged only on interfaces. A static abstract method on an abstract class is illegal and we shouldn't tag such cases.
+                if (runtimeFeatureType != null && ProcessPreviewAttribute(runtimeFeatureType, requiresPreviewFeaturesSymbols))
+                {
+                    requiresPreviewFeaturesSymbolsToUsageType.GetOrAdd(propertyOrMethodSymbol, PreviewFeatureUsageType.StaticAbstract);
+                    return true;
+                }
+            }
+
+            if (propertyOrMethodSymbol.IsImplementationOfAnyImplicitInterfaceMember(out ISymbol baseInterfaceMember))
+            {
+                if (ProcessPreviewAttribute(baseInterfaceMember, requiresPreviewFeaturesSymbols) || ContainingTypeIsPreview(baseInterfaceMember, requiresPreviewFeaturesSymbols))
+                {
+                    return true;
+                }
+            }
+
+            if (propertyOrMethodSymbol.IsOverride)
+            {
+                ISymbol? overridden = propertyOrMethodSymbol.GetOverriddenMember();
+                if (overridden != null && ProcessPreviewAttribute(overridden, requiresPreviewFeaturesSymbols))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
+        {
+            ISymbol symbol = context.Symbol;
+
+            // Don't report diagnostics on types that are marked as Preview
+            if (ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols))
```

Comment:
Wondering what makes sure that it was a type symbol, seems this analyses many different symbols `SymbolKind.NamedType, SymbolKind.Method, SymbolKind.Property, SymbolKind.Field, SymbolKind.Event`

Reply:
This was just a bad comment :) I meant everything when I wrote "types" here. Deleted the comment now. For ex: It's ok to see the `IMethodSymbol` for a preview method here. We will just add the symbol to the dictionary and return without reporting a diagnostic.